Phase 2 of the multi-threaded framework

Dr15Jones opened the pull request:
This completes phase 2 of the multi-threaded framework. In this phase event processing is handled by a series of 'chained' tbb tasks. This allows maximal use of multiple threads for each event.

Dr15Jones committed:
Removed unnecessary argument from processOneEvent

The argument cleaningUpAfterException had no function within the
processOneEvent method.

Dr15Jones committed:
Have StreamSchedule directly call WorkerManager::setupOnDemandSystem

The StreamSchedule uses a specialized call to process events so there
is no longer a reason to call the general WorkerManager::processOneOccurence
and have that function have special compile time switches to handle events.

Dr15Jones committed:
Added WaitingTaskHolder class

The WaitingTaskHolder manages one task, compared to WaitingTaskList
which manages many tasks.

Dr15Jones committed:
Break StreamSchedule event processing into multiple tasks

Add processOneEventAsync method which internally breaks the processing
of the event into multiple tasks
   -setup event and process paths
   -get TriggerResults and then process end_paths
   -finalize event

Dr15Jones committed:
Move looper handling to its own function

Within EventProcessor, moved the interactions with the looper
during event processing into its own function.

Dr15Jones committed:
Added default constructor to WaitingTaskHolder

Dr15Jones committed:
Properly handle activating Services in threads

Make sure that the Service token is propagated to tasks in
StreamSchedule.

Dr15Jones committed:
Move StreamSchedule::processOneEventAsyc to public

Need to be able to call StreamSchedule::processOneEventAsync from
Schedule.

Dr15Jones committed:
Asynchronously process one event

Asynchronously handle all aspects of processing one event. This
excludes reading the event from the Source.

Dr15Jones committed:
Update test results to match asynchronous processing of events

The order in which modules are run changed slightly due to the
asynchronous handling of multiple SubProcesses.

Dr15Jones committed:
Regularize the handling of the first event in a lumi

The EventProcessor has to treat the first event in a lumi in a
special way because it can not ask the source what the next transition
is again since it would advance to the next transition. Previously
the code to handle this was broken into two different areas and
required the event to be handled by stream 0. This change makes the
handling of the first event closer to how other events are handled.

This required updating one unit test comparison file because the
module processing order is changed in a non-important way.

Dr15Jones committed:
Implement full asynchronous interface to EventProcessor

Implemented full asynchronous processing of modules within events
through the whole system. Access to the source is now protected by
a SerialTaskQueue instead of a mutex.

The exact ordering of module calling was changed which required
updating one unit test output comparison file.

Dr15Jones committed:
Code cleanup after asynchronous event feature

Removed functions which were replaced with asynchronous
implementations. Several other small changes.

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for CMSSW_9_0_X.

It involves the following packages:

FWCore/Concurrency
FWCore/Framework
FWCore/Integration


@cmsbuild, @smuzaffar, @Dr15Jones, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit, @wmtan this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
@knoepfel @wddgit please look over this change and make any comments

cmsbuild commented:
-1

Tested at: 757d039be085938a689d24908b8ae8783fba3256

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16845/16747/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestIntegrationUnscheduledFailOnOutput had ERRORS
---> test TestIntegrationParameterSet had ERRORS

cmsbuild commented:
Comparison job queued.

Dr15Jones commented:
The unit test failures are unrelated to this pull request and point to a problem with python

```
python encountered the error: <type 'exceptions.NameError'>
name 'os' is not defined
```

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16845/16747/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar

cmsbuild commented:
-1

Tested at: 757d039be085938a689d24908b8ae8783fba3256

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16845/16750/summary.html

I found follow errors while testing this PR

**Failed tests:  AddOn**

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,HLT:@relval2016,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 --magField 38T_PostLS1 : FAILED - time: date Sat Dec  3 18:05:37 2016-date Sat Dec  3 17:54:50 2016 s - exit: 34304

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16845/16750/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

Dr15Jones commented:
-1

Dr15Jones commented:
The test failure seems to point to the case of the Event being reset while the PoolOutputModule is running.

Dr15Jones commented:
@smuzaffar @davidlange6 can we open an CMSSW_9_0_DEVEL_X branch and put this pull request there? It would help me alot to be able to have a full release to do more extensive testing.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
I've rerun addOnTests myself and they all pass. I then ran the job which had failed during testing a half dozen times and it has always passed. I'm now running valgrind to see if there might have been a memory problem.

cmsbuild commented:
-1

Tested at: 757d039be085938a689d24908b8ae8783fba3256

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16845/16754/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestFWCoreFrameworkGlobalStreamOne had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16845/16754/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

Dr15Jones commented:
My valgrind run shows a large number of invalid writes coming from other code. This could very well have caused the original test failure. I'll be creating an issue for the problem discovered.

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs (but tests are reportedly failing). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar

Dr15Jones commented:
The unit test failure does not point a problem with this pull request. However, it does show a problem with the test because the framework change (which is an improvement) is more capable of scheduling modules to run concurrently than the test thinks should be. I will fix the unit test in a different pull request.

smuzaffar commented:
@Dr15Jones , CMSSW_9_0_DEVEL_X branch is now available

knoepfel commented on FWCore/Concurrency/interface/WaitingTaskHolder.h:
```diff
@@ -0,0 +1,84 @@
+  public:
+    WaitingTaskHolder():
+    m_task(nullptr) {}
+    
```
Somewhat a matter of taste, but initializing `m_task` to `nullptr` in-class allows you to just specify:

    WaitingTaskHolder() = default;

Of course, doing so will not save you much maintenance for this class since there's only one data member.  But extending the class in the future is then easier.

knoepfel replied:
You are correct that providing `= default` is equivalent (in behavior) to specifying `WaitingTaskHolder(){}`.  However, when in-class initialization is used, the member is still initialized to its user-provided, in-class default value, even if an initialization list is absent in the c'tor.  If an in-class, default-initialized member is also specified in the c'tor's initialization list, the c'tor initialization takes precedence.

knoepfel commented on FWCore/Framework/interface/EventProcessor.h:
```diff
@@ -285,7 +295,8 @@ namespace edm {
     
-    std::mutex                                    nextTransitionMutex_;
+    SharedResourcesAcquirer                       sourceResourcesAcquirer_;
+    std::shared_ptr<std::recursive_mutex>         sourceMutex_;
```
Can `sourceMutex_` be a `unique_ptr`?

Dr15Jones replied:
It is shared with the DelayedReader so must be `std::shared_ptr`.

knoepfel commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -554,92 +556,146 @@ namespace edm {
+                                              ptr = *iPtr;
+                                            }
+                                            finishedPaths(ptr, std::move(iTask), ep, es);
+                                          });
```
Is the lambda marked `mutable` only because of the `std::move(iTask)`?  If so, is it possible to pass `iTask` as an r-value reference function argument so as to avoid the `mutable` decorator, or are you restricted by TBB's interface?

Dr15Jones replied:
It is mutable because of iTask. I can't change the function signature of the lambda, hence why I have to capture the local variable.

knoepfel commented on FWCore/Framework/src/SubProcess.cc:
```diff
@@ -327,20 +332,21 @@ namespace edm {
     EventAuxiliary aux(principal.aux());
     aux.setProcessHistoryID(principal.processHistoryID());
-
+    
```
Is it possible to have your editor removing spurious whitespaces like this?  If not, it's just a few extra bytes....

knoepfel commented on FWCore/Framework/src/StreamSchedule.cc:
```diff
@@ -554,92 +556,146 @@ namespace edm {
+    for(auto it = trig_paths_.rbegin(), itEnd = trig_paths_.rend();
+        it != itEnd; ++ it) {
+      it->processOneOccurrenceAsync(pathsDone,ep, es, streamID_, &streamContext_);
+    }
```
Might consider using `boost/range/adaptor/reversed.hpp` here as you do below.

knoepfel commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -1858,130 +1870,141 @@ namespace edm {
+      //need to use lock in addition to the serial task queue because
+      // of delayed provenance reading and reading data in response to
+      // edm::Refs etc
+      std::lock_guard<std::recursive_mutex> guard(*(sourceMutex_.get()));
```
Using `*sourceMutex_` is equivalent and more succinct.

Dr15Jones replied:
good point.

Dr15Jones committed:
Update SecondaryEventProvider to match framework change

The phase 2 change of the threaded framework no longer uses
WorkerManger::processOneOccurrence for handling Events. The
SecondaryEventProvider now uses the setupUnscheduledSystem call
just like the framework.

cmsbuild commented:
Pull request #16845 was updated. @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @cmsbuild, @davidlange6 can you please check and sign again.

Dr15Jones committed:
Make sure loop completes before callback task runs

When processing the end paths the code was not guaranteeing that
the loop completes before the endPathDone task can start. This
did happen when there were many end paths in the job where some
ended very quickly and the thread with the loop happened to have
been paused by the system.

cmsbuild commented:
Pull request #16845 was updated. @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @cmsbuild, @davidlange6 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
@davidlange6 now that CMSSW_9_0_0_pre2 and CMSSW_8_2_0 are out, when can we shedule this pull request to be merged?

Dr15Jones commented:
+1

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar

cmsbuild commented:
-1

Tested at: c80ed8a7a6be4a91db78e551ccb4353f6d5510ac

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16845/17067/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test DetectorDescriptionRegressionTestDOMCount had ERRORS

cmsbuild commented:
Comparison job queued.

Dr15Jones commented:
The test failure is occuring in the IB used for the testing. Therefore it is unrelated to this pull request.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16845/17067/summary.html

Alternative comparison was/were failed for workflow(s):
25202.0
25.0
1306.0
21234.0
1003.0
23234.0
140.53
136.731
1330.0
1000.0
135.4
10024.0
5.1
8.0
1001.0
10021.0
50202.0
20034.0
9.0
4.22
